File: cpp/initialize-agent-with-aes-persistor/src/InitializeAgentWithAesPersistor.cpp
/*
 * (c) 2018 Ionic Security Inc.
 * By using this code, I agree to the Terms & Conditions (https://dev.ionic.com/use.html)
 * and the Privacy Policy (https://www.ionic.com/privacy-notice/).
 */

#include "ISAgent.h"
#include "ISAgentSDKError.h"
#include <stdio.h>
#include <cstdlib>
#include <iostream>

int main(int argc, char* argv[]) {

    int nErrorCode;
    std::string persistorPath = "../../sample-data/persistors/sample-persistor.aes";
    std::string persistorKeyHex = "A0444B8B5A7209780823617A98986831B8240BAA851A0B1696B0329280286B17";
    std::string persistorAuthDataStr = "persistor auth data";

    // initialize agent with aes persistor
    ISAgent agent;
    ISAgentDeviceProfilePersistorAesGcm persistor;
    persistor.setFilePath(persistorPath);
    ISCryptoBytes persistorKey;
    ISCryptoHexString persistorKeyHexStr(persistorKeyHex);
    persistorKeyHexStr.toBytes(persistorKey);
    persistor.setKey(persistorKey);
    ISCryptoBytes persistorAuthData((byte*)persistorAuthDataStr.data(), persistorAuthDataStr.size());
    persistor.setAuthData(persistorAuthData);
    nErrorCode = agent.initialize(persistor);
    if (nErrorCode != ISAGENT_OK) {
        std::cerr << "Error initializing agent: " << ISAgentSDKError::getErrorCodeString(nErrorCode) << std::endl;
    }

    // display all profiles in persistor
    const std::vector<ISAgentDeviceProfile> profiles = agent.getAllProfiles();
    for (int i=0; i < profiles.size(); i++) {
        const ISAgentDeviceProfile profile = profiles.at(i);
        std::cout << "---" << std::endl;
        std::cout << "ID       : " << profile.getDeviceId() << std::endl;
        std::cout << "Name     : " << profile.getName() << std::endl;
        std::cout << "Keyspace : " << profile.getKeySpace() << std::endl;
        std::cout << "ApiUrl   : " << profile.getServer() << std::endl;
    }
}

File: java/initialize-agent-with-aes-persistor/src/main/java/com/ionic/samples/InitializeAgentWithAesPersistor.java
/*
 * (c) 2018 Ionic Security Inc.
 * By using this code, I agree to the Terms & Conditions (https://dev.ionic.com/use.html)
 * and the Privacy Policy (https://www.ionic.com/privacy-notice/).
 * 
 * Developed with Ionic Java SDK 2.1.0
 */

package com.ionic.samples;

import com.ionic.sdk.agent.Agent;
import com.ionic.sdk.device.profile.DeviceProfile;
import com.ionic.sdk.device.profile.persistor.DeviceProfilePersistorAesGcm;
import com.ionic.sdk.error.IonicException;
import java.util.List;
import javax.xml.bind.DatatypeConverter;

public class InitializeAgentWithAesPersistor
{
    public static void main(String[] args)
    {
        String persistorPath = "../../sample-data/persistors/sample-persistor.aes";
        byte[] persistorKey = DatatypeConverter.parseHexBinary("A0444B8B5A7209780823617A98986831B8240BAA851A0B1696B0329280286B17");
        byte[] persistorAuthData = "persistor auth data".getBytes();

        // initialize agent with aes persistor        
        Agent agent = new Agent();
        try {
            DeviceProfilePersistorAesGcm persistor = new DeviceProfilePersistorAesGcm(persistorPath);
            persistor.setKey(persistorKey);
            persistor.setAuthData(persistorAuthData);
            agent.initialize(persistor);
        } catch(IonicException e) {
            System.out.println(e.getMessage());
            System.exit(1);
        }

        // display all profiles in persistor
        List<DeviceProfile> profiles = agent.getAllProfiles();
        for (DeviceProfile profile : profiles) {
            System.out.println("---");
            System.out.println("Id       : " + profile.getDeviceId());
            System.out.println("Name     : " + profile.getName());
            System.out.println("Keyspace : " + profile.getKeySpace());
            System.out.println("ApiUrl   : " + profile.getServer());
        }
    }
}
